### PR TITLE
Add tests to serializable operations validation

### PR DESCRIPTION
**Pull request by serathius**
cc @fuweid @siyuanfoundation @MadhavJivrajani

### CODE REVIEW

**Comment by fuweid** [comment]
Hmm, it's timeout error

File: tests/robustness/validate/operations.go [hunk 1]
```diff
@@ -52,45 +50,50 @@ func validateLinearizableOperationsAndVisualize(lg *zap.Logger, operations []por
...
+		return fmt.Errorf("request about a future rev with response")
 	}
 
 	_, expectResp := state.Step(request)
-	if !reflect.DeepEqual(response.EtcdResponse.Range, expectResp.Range) {
-		t.Errorf("Invalid serializable response, diff: %s", cmp.Diff(response.EtcdResponse.Range, expectResp.Range))
+
+	if diff := cmp.Diff(response.EtcdResponse.Range, expectResp.Range); diff != "" {
+		lg.Error("Failed validating serializable operation", zap.Any("request", request), zap.String("diff", diff))
+		return fmt.Errorf("response didn't match expected")
```
Comment (siyuanfoundation) [inline tests/robustness/validate/operations.go:96]: make the error a constant so it is easier to check in tests.

File: tests/robustness/validate/operations_test.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+					Output: rangeResponse(3,
+						keyValue("a", "1", 2),
+						keyValue("b", "2", 3),
+						keyValue("c", "3", 4),
+					),
+				},
+				{
+					Input: rangeRequest("a", "z", 4, 2),
+					Output: rangeResponse(3,
+						keyValue("a", "1", 2),
```
Comment (siyuanfoundation) [inline tests/robustness/validate/operations_test.go:85]: what happens if the order of the keys are reversed? Maybe add a test case for it.

File: tests/robustness/validate/operations_test.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			replay := model.NewReplay(tc.persistedRequests)
+			err := validateSerializableOperations(zaptest.NewLogger(t), tc.operations, replay)
+			var errStr string
+			if err != nil {
+				errStr = err.Error()
+			}
+			if errStr != tc.expectError {
+				t.Errorf("validateWatch(...), got: %q, want: %q", err, tc.expectError)
```
Comment (MadhavJivrajani) [inline tests/robustness/validate/operations_test.go:228]:
```suggestion
				t.Errorf("validateSerializableOperations(...), got: %q, want: %q", err, tc.expectError)
```

File: tests/robustness/validate/operations.go [hunk 4]
```diff
@@ -52,45 +50,50 @@ func validateLinearizableOperationsAndVisualize(lg *zap.Logger, operations []por
...
-	for _, op := range operations {
-		request := op.Input.(model.EtcdRequest)
-		if request.Type == model.Range && request.Range.Revision != 0 {
-			resp = append(resp, op)
+	for _, client := range clients {
+		for _, op := range client.KeyValue {
+			request := op.Input.(model.EtcdRequest)
+			if request.Type == model.Range && request.Range.Revision != 0 {
+				resp = append(resp, op)
+			}
```
Comment (MadhavJivrajani) [inline tests/robustness/validate/operations.go:73]: Just a question: iiuc, all range requests are linearized by default, so:
(1) how are we doing serialized ranges? I see how we implement them in our model, but not sure of the distinction.
(2) how is this `if` condition telling us that an operation is a serializable operation?

File: tests/robustness/validate/operations.go (hunk 1 above)
Comment (serathius) [reply tests/robustness/validate/operations.go:96]: That's not a huge priority. Feel free to open a followup PR.

File: tests/robustness/validate/operations_test.go (hunk 2 above)
Comment (serathius) [reply tests/robustness/validate/operations_test.go:85]: Right!

File: tests/robustness/validate/operations_test.go (hunk 3 above)
Comment (serathius) [reply tests/robustness/validate/operations_test.go:228]: Done

**Review by fuweid — APPROVED** [review]
LGTM

**Review by siyuanfoundation — APPROVED** [review]

**Comment by serathius** [comment]
ping @MadhavJivrajani

**Comment by serathius** [comment]
@MadhavJivrajani is on vacation.

**Pull request merged**